llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1076,15 +1078,39 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
       TypeInfo.inferScalarType(R.getOperand(1)) ==
           TypeInfo.inferScalarType(R.getVPSingleValue()))
     return R.getVPSingleValue()->replaceAllUsesWith(R.getOperand(1));
+
+  // Constant folding. TODO: cast and cmp.
+  VPConstantFolder Folder(DL);
+  VPlan *Plan = R.getParent()->getPlan();
+  if (match(&R, m_BinaryAnd(m_VPValue(X), m_VPValue(Y))))
+    if (Value *V = Folder.foldAnd(X, Y))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
+  if (match(&R, m_BinaryOr(m_VPValue(X), m_VPValue(Y))))
+    if (Value *V = Folder.foldOr(X, Y))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
+  if (match(&R, m_Not(m_VPValue(X))))
+    if (Value *V = Folder.foldNot(X))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
+  if (match(&R, m_LogicalAnd(m_VPValue(X), m_VPValue(Y))))
+    if (Value *V = Folder.foldLogicalAnd(X, Y))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
+  if (match(&R, m_Select(m_VPValue(X), m_VPValue(Y), m_VPValue(Z))))
+    if (Value *V = Folder.foldSelect(X, Y, Z))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
+  if (match(&R, m_GetElementPtr(m_VPValue(X), m_VPValue(Y))))
+    if (Value *V = Folder.foldPtrAdd(
+            X, Y, cast<VPRecipeWithIRFlags>(R).getGEPNoWrapFlags()))
+      R.getVPSingleValue()->replaceAllUsesWith(Plan->getOrAddLiveIn(V));
```

Comment:
Would it be possible to have a single `tryToConstantFold(unsigned Opcode, ArrayRef<VPValue *> Ops)` that calls the corresponding `Folder.foldXXX` functions? Then there would only be a single place where we need to add folding support for new opcodes?

The relevant recipe classes should provide a `::getOpcode()` helper, which is used by the constant folder. Maybe it would be possible to use that?

Reply:
Just to clarify: what should I call tryToConstantFold with? `R->getOpcode()`, and is there some API for getting the ops in a generic fashion? I wasn't able to find this, which is why I used the recipe matchers.

Reply:
I think it would need something like a type switch to cast to specific recipe types that have `getOpcode` and operands can be accessed by `operands()`.